File: TwoWayDatabinding/app/src/main/java/com/enpassio/twowaydatabinding/data/ToyRepository.java
package com.enpassio.twowaydatabinding.data;

import android.arch.lifecycle.LiveData;

import com.enpassio.twowaydatabinding.data.model.ToyEntry;
import com.enpassio.twowaydatabinding.utils.AppExecutors;

import java.util.List;

public class ToyRepository {

    private static ToyRepository sInstance;
    private final ToyDatabase mDatabase;
    private final AppExecutors mExecutors;

    private ToyRepository(ToyDatabase database, AppExecutors executors) {
        mDatabase = database;
        mExecutors = executors;
    }

    public static ToyRepository getInstance(ToyDatabase database, AppExecutors executors) {
        if (sInstance == null) {
            synchronized (ToyRepository.class) {
                if (sInstance == null) {
                    sInstance = new ToyRepository(database, executors);
                }
            }
        }
        return sInstance;
    }

    public LiveData<List<ToyEntry>> getToyList() {
        return mDatabase.toyDao().getAllToys();
    }

    public LiveData<ToyEntry> getChosenToy(int toyId){
        return mDatabase.toyDao().getChosenToy(toyId);
    }

    public void insertToy(final ToyEntry toy){
        mExecutors.diskIO().execute(() -> mDatabase.toyDao().insertToy(toy));
    }

    public void updateToy(final ToyEntry toy){
        mExecutors.diskIO().execute(() -> mDatabase.toyDao().updateToyInfo(toy));
    }

    public void deleteToy(final ToyEntry toy){
        mExecutors.diskIO().execute(() -> mDatabase.toyDao().deleteToy(toy));
    }

}

File: TwoWayDatabindingKotlin/app/src/main/java/com/enpassion/twowaydatabindingkotlin/ToyRepository.kt
package com.enpassion.twowaydatabindingkotlin

import android.arch.lifecycle.LiveData
import com.enpassion.twowaydatabindingkotlin.data.ToyDatabase
import com.enpassion.twowaydatabindingkotlin.data.ToyEntry
import com.enpassion.twowaydatabindingkotlin.utils.AppExecutors


class ToyRepository private constructor(private val mDatabase: ToyDatabase, private val mExecutors: AppExecutors) {

    val toyList: LiveData<List<ToyEntry>>
        get() = mDatabase.toyDao().allToys

    fun getChosenToy(toyId: Int): LiveData<ToyEntry> {
        return mDatabase.toyDao().getChosenToy(toyId)
    }

    fun insertToy(toy: ToyEntry) {
        mExecutors.diskIO().execute{ mDatabase.toyDao().insertToy(toy) }
    }

    fun updateToy(toy: ToyEntry) {
        mExecutors.diskIO().execute{ mDatabase.toyDao().updateToyInfo(toy) }
    }

    fun deleteToy(toy: ToyEntry) {
        mExecutors.diskIO().execute{ mDatabase.toyDao().deleteToy(toy) }
    }

    companion object {

        private var sInstance: ToyRepository? = null

        fun getInstance(database: ToyDatabase, executors: AppExecutors): ToyRepository {
            return sInstance ?: synchronized(this) {
                sInstance ?: ToyRepository(database, executors).also { sInstance = it }
            }
        }
    }

}
